File: webapp/scripts/parse-recipes.ts
/**
 * parse-recipes.ts
 * 
 * Parses DTRecipeProvider.java and extracts all crafting recipes into a JSON file.
 * Run with: npx tsx scripts/parse-recipes.ts
 */

import * as fs from 'fs';
import * as path from 'path';

interface ShapedRecipe {
  type: 'shaped';
  pattern: string[];
  key: { [char: string]: string };
  output: { item: string; count: number };
}

interface ShapelessRecipe {
  type: 'shapeless';
  ingredients: string[];
  output: { item: string; count: number };
}

interface CookingRecipe {
  type: 'smelting' | 'campfire' | 'smoking';
  input: string;
  output: { item: string; count: number };
}

type Recipe = ShapedRecipe | ShapelessRecipe | CookingRecipe;

// Map Java item references to namespaced IDs
function resolveItemId(javaRef: string): string {
  javaRef = javaRef.trim();
  
  // Vanilla Items.*
  if (javaRef.startsWith('Items.')) {
    const name = javaRef.replace('Items.', '').toLowerCase();
    return `minecraft:${name}`;
  }
  // Vanilla Blocks.*
  if (javaRef.startsWith('Blocks.')) {
    const name = javaRef.replace('Blocks.', '').toLowerCase();
    return `minecraft:${name}`;
  }
  // DoggyItems.*
  if (javaRef.startsWith('DoggyItems.')) {
    let name = javaRef.replace('DoggyItems.', '').replace('.get()', '').toLowerCase();
    return `doggytalents:${name}`;
  }
  // DoggyBlocks.*
  if (javaRef.startsWith('DoggyBlocks.')) {
    let name = javaRef.replace('DoggyBlocks.', '').replace('.get()', '').toLowerCase();
    return `doggytalents:${name}`;
  }
  // ItemTags.*
  if (javaRef.startsWith('ItemTags.')) {
    const name = javaRef.replace('ItemTags.', '').toLowerCase();
    return `tag:${name}`;
  }
  // BlockTags.*
  if (javaRef.startsWith('BlockTags.')) {
    const name = javaRef.replace('BlockTags.', '').toLowerCase();
    return `tag:${name}`;
  }
  
  return `unknown:${javaRef}`;
}

function parseRecipes(javaSource: string): { [id: string]: Recipe } {
  const recipes: { [id: string]: Recipe } = {};
  
  // ===== SHAPED RECIPES =====
  // Match ShapedRecipeBuilder blocks
  const shapedRegex = /ShapedRecipeBuilder\.shaped\s*\(\s*RecipeCategory\.\w+,\s*([\w.()]+?)(?:\.get\(\))?\s*(?:,\s*(\d+))?\s*\)([\s\S]*?)\.save\s*\(\s*consumer\s*(?:,\s*[^)]+)?\s*\)/g;
  
  let match;
  while ((match = shapedRegex.exec(javaSource)) !== null) {
    const outputRef = match[1];
    const count = parseInt(match[2] || '1');
    const body = match[3];
    
    // Extract patterns
    const patternMatches = [...body.matchAll(/\.pattern\s*\(\s*"([^"]{1,3})"\s*\)/g)];
    const pattern = patternMatches.map(m => m[1]);
    
    // Pad patterns to 3 chars
    const paddedPattern = pattern.map(p => p.padEnd(3, ' '));
    
    // Extract key definitions
    const key: { [char: string]: string } = {};
    const defineMatches = [...body.matchAll(/\.define\s*\(\s*'(\w)'\s*,\s*([\w.()]+(?:\.get\(\))?)\s*\)/g)];
    for (const dm of defineMatches) {
      key[dm[1]] = resolveItemId(dm[2]);
    }
    
    // Generate recipe ID from output
    const outputItem = resolveItemId(outputRef);
    const recipeId = outputItem.split(':')[1] || outputItem;
    
    // Check if we already have this recipe (alt recipes)
    let finalId = recipeId;
    if (recipes[finalId]) {
      // Check for explicit alt name in save()
      const saveMatch = match[0].match(/save\s*\(\s*consumer\s*,\s*Util\.getResource\s*\(\s*"([^"]+)"\s*\)\s*\)/);
      if (saveMatch) {
        finalId = saveMatch[1];
      } else {
        finalId = `${recipeId}_alt`;
      }
    }
    
    recipes[finalId] = {
      type: 'shaped',
      pattern: paddedPattern,
      key,
      output: { item: outputItem, count }
    };
  }
  
  // ===== SHAPELESS RECIPES =====
  const shapelessRegex = /ShapelessRecipeBuilder\.shapeless\s*\(\s*RecipeCategory\.\w+,\s*([\w.()]+?)(?:\.get\(\))?\s*(?:,\s*(\d+))?\s*\)([\s\S]*?)\.save\s*\(\s*consumer\s*(?:,\s*[^)]+)?\s*\)/g;
  
  while ((match = shapelessRegex.exec(javaSource)) !== null) {
    const outputRef = match[1];
    const count = parseInt(match[2] || '1');
    const body = match[3];
    
    // Extract ingredients
    const ingredients: string[] = [];
    const reqMatches = [...body.matchAll(/\.requires\s*\(\s*([\w.()]+(?:\.get\(\))?)\s*(?:,\s*(\d+))?\s*\)/g)];
    for (const rm of reqMatches) {
      const itemId = resolveItemId(rm[1]);
      const qty = parseInt(rm[2] || '1');
      for (let i = 0; i < qty; i++) {
        ingredients.push(itemId);
      }
    }
    
    const outputItem = resolveItemId(outputRef);
    const recipeId = outputItem.split(':')[1] || outputItem;
    
    let finalId = recipeId;
    if (recipes[finalId]) {
      const saveMatch = match[0].match(/save\s*\(\s*consumer\s*,\s*Util\.getResource\s*\(\s*"([^"]+)"\s*\)\s*\)/);
      if (saveMatch) {
        finalId = saveMatch[1];
      } else {
        finalId = `${recipeId}_shapeless`;
      }
    }
    
    recipes[finalId] = {
      type: 'shapeless',
      ingredients,
      output: { item: outputItem, count }
    };
  }
  
  // ===== COOKING RECIPES (via registerTripleCooking) =====
  const cookingRegex = /registerTripleCooking\s*\(\s*consumer\s*,\s*\n?\s*Ingredient\.of\s*\(\s*([\w.()]+(?:\.get\(\))?)\s*\)\s*,\s*\n?\s*([\w.()]+(?:\.get\(\))?)\s*,/g;
  
  while ((match = cookingRegex.exec(javaSource)) !== null) {
    const inputItem = resolveItemId(match[1]);
    const outputItem = resolveItemId(match[2]);
    const recipeId = (outputItem.split(':')[1] || outputItem);
    
    recipes[`${recipeId}_smelting`] = {
      type: 'smelting',
      input: inputItem,
      output: { item: outputItem, count: 1 }
    };
    
    recipes[`${recipeId}_campfire`] = {
      type: 'campfire',
      input: inputItem,
      output: { item: outputItem, count: 1 }
    };
    
    recipes[`${recipeId}_smoking`] = {
      type: 'smoking',
      input: inputItem,
      output: { item: outputItem, count: 1 }
    };
  }
  
  return recipes;
}

// Main
const javaPath = path.join(__dirname, '..', 'DTRecipeProvider.java');
const outputDir = path.join(__dirname, '..', 'data');
const outputPath = path.join(outputDir, 'recipes.json');

if (!fs.existsSync(javaPath)) {
  console.error(`Java file not found: ${javaPath}`);
  process.exit(1);
}

const javaSource = fs.readFileSync(javaPath, 'utf8');
const recipes = parseRecipes(javaSource);

if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

fs.writeFileSync(outputPath, JSON.stringify(recipes, null, 2));

console.log(`Parsed ${Object.keys(recipes).length} recipes → ${outputPath}`);

// Print summary
const types = { shaped: 0, shapeless: 0, smelting: 0, campfire: 0, smoking: 0 };
for (const r of Object.values(recipes)) {
  types[r.type] = (types[r.type] || 0) + 1;
}
console.log('Recipe types:', types);

File: webapp/scripts/parse_recipes.py
import json
import os
import re

JAVA_PATH = os.path.join(os.path.dirname(__file__), '../ref/DTRecipeProvider.java')
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), '../data')
OUTPUT_PATH = os.path.join(OUTPUT_DIR, 'recipes.json')

def resolve_item_id(java_ref):
    java_ref = java_ref.strip()
    
    # Vanilla Items.* or Blocks.*
    if java_ref.startswith('Items.'):
        name = java_ref.replace('Items.', '').lower()
        return f"minecraft:{name}"
    if java_ref.startswith('Blocks.'):
        name = java_ref.replace('Blocks.', '').lower()
        return f"minecraft:{name}"
        
    # DoggyItems.* or DoggyBlocks.*
    if java_ref.startswith('DoggyItems.'):
        name = java_ref.replace('DoggyItems.', '').replace('.get()', '').lower()
        return f"doggytalents:{name}"
    if java_ref.startswith('DoggyBlocks.'):
        name = java_ref.replace('DoggyBlocks.', '').replace('.get()', '').lower()
        return f"doggytalents:{name}"
        
    # Tags
    if java_ref.startswith('ItemTags.'):
        name = java_ref.replace('ItemTags.', '').lower()
        return f"tag:{name}"
    if java_ref.startswith('BlockTags.'):
        name = java_ref.replace('BlockTags.', '').lower()
        return f"tag:{name}"
        
    return f"unknown:{java_ref}"

def parse_recipes(java_source):
    recipes = {}
    
    # ===== SHAPED RECIPES =====
    shaped_regex = r'ShapedRecipeBuilder\.shaped\s*\(\s*RecipeCategory\.\w+,\s*([\w.()]+?)(?:\.get\(\))?\s*(?:,\s*(\d+))?\s*\)([\s\S]*?)\.save\s*\(\s*consumer\s*(?:,\s*[^)]+)?\s*\)'
    for match in re.finditer(shaped_regex, java_source):
        output_ref = match.group(1)
        count = int(match.group(2) or '1')
        body = match.group(3)
        
        # Patterns
        patterns = re.findall(r'\.pattern\s*\(\s*"([^"]{1,3})"\s*\)', body)
        padded_patterns = [p.ljust(3) for p in patterns]
        
        # Keys
        key = {}
        for char, ref in re.findall(r'\.define\s*\(\s*\'(\w)\'\s*,\s*([\w.()]+(?:\.get\(\))?)\s*\)', body):
            key[char] = resolve_item_id(ref)
            
        output_item = resolve_item_id(output_ref)
        recipe_id = output_item.split(':')[1] if ':' in output_item else output_item
        
        final_id = recipe_id
        if final_id in recipes:
            save_match = re.search(r'save\s*\(\s*consumer\s*,\s*Util\.getResource\s*\(\s*"([^"]+)"\s*\)\s*\)', match.group(0))
            if save_match:
                final_id = save_match.group(1)
            else:
                final_id = f"{recipe_id}_alt"
        
        recipes[final_id] = {
            "type": "shaped",
            "pattern": padded_patterns,
            "key": key,
            "output": {"item": output_item, "count": count}
        }

    # ===== SHAPELESS RECIPES =====
    shapeless_regex = r'ShapelessRecipeBuilder\.shapeless\s*\(\s*RecipeCategory\.\w+,\s*([\w.()]+?)(?:\.get\(\))?\s*(?:,\s*(\d+))?\s*\)([\s\S]*?)\.save\s*\(\s*consumer\s*(?:,\s*[^)]+)?\s*\)'
    for match in re.finditer(shapeless_regex, java_source):
        output_ref = match.group(1)
        count = int(match.group(2) or '1')
        body = match.group(3)
        
        ingredients = []
        for ref, qty in re.findall(r'\.requires\s*\(\s*([\w.()]+(?:\.get\(\))?)\s*(?:,\s*(\d+))?\s*\)', body):
            item_id = resolve_item_id(ref)
            count_ing = int(qty or '1')
            for _ in range(count_ing):
                ingredients.append(item_id)
                
        output_item = resolve_item_id(output_ref)
        recipe_id = output_item.split(':')[1] if ':' in output_item else output_item
        
        final_id = recipe_id
        if final_id in recipes:
            save_match = re.search(r'save\s*\(\s*consumer\s*,\s*Util\.getResource\s*\(\s*"([^"]+)"\s*\)\s*\)', match.group(0))
            if save_match:
                final_id = save_match.group(1)
            else:
                final_id = f"{recipe_id}_shapeless"
                
        recipes[final_id] = {
            "type": "shapeless",
            "ingredients": ingredients,
            "output": {"item": output_item, "count": count}
        }

    # ===== COOKING RECIPES =====
    cooking_regex = r'registerTripleCooking\s*\(\s*consumer\s*,\s*\n?\s*Ingredient\.of\s*\(\s*([\w.()]+(?:\.get\(\))?)\s*\)\s*,\s*\n?\s*([\w.()]+(?:\.get\(\))?)\s*,'
    for match in re.finditer(cooking_regex, java_source):
        input_item = resolve_item_id(match.group(1))
        output_item = resolve_item_id(match.group(2))
        recipe_id = output_item.split(':')[1] if ':' in output_item else output_item
        
        for suffix in ['_smelting', '_campfire', '_smoking']:
            recipes[f"{recipe_id}{suffix}"] = {
                "type": suffix[1:],
                "input": input_item,
                "output": {"item": output_item, "count": 1}
            }
            
    return recipes

def main():
    if not os.path.exists(JAVA_PATH):
        print(f"Error: Java file not found at {JAVA_PATH}")
        return

    with open(JAVA_PATH, 'r') as f:
        java_source = f.read()
        
    recipes = parse_recipes(java_source)
    
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)
        
    with open(OUTPUT_PATH, 'w') as f:
        json.dump(recipes, f, indent=2)
        
    print(f"Parsed {len(recipes)} recipes -> {OUTPUT_PATH}")
    
    # Summary
    types = {}
    for r in recipes.values():
        t = r['type']
        types[t] = types.get(t, 0) + 1
    print("Recipe types:", types)

if __name__ == "__main__":
    main()
